File: backend/src/main/java/com/example/contenthub/service/crawling/novel/KakaoPageCrawler.java
package com.example.contenthub.service.crawling.novel;

import com.example.contenthub.constants.crawling.KakaoNovelXPath;
import com.example.contenthub.constants.crawling.NaverNovelXPath;
import com.example.contenthub.dto.ContentCrawlDTO;
import com.example.contenthub.entity.Content;
import com.example.contenthub.exception.CrawlerException;
import com.example.contenthub.service.auth.social.KakaoPageLogin;
import com.google.gson.Gson;
import lombok.RequiredArgsConstructor;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Component
@RequiredArgsConstructor
public class KakaoPageCrawler {
    private final KakaoPageLogin kakaoLogin;

    public List<ContentCrawlDTO> crawl(WebDriver mainDriver) {
        List<ContentCrawlDTO> novels = new ArrayList<>();
        try {
            setUpDrivers(mainDriver);

            scrollPageToBottom(mainDriver);

            List<String> novelLinks = getKakaoPageData(mainDriver);
            for (String href : novelLinks) {
                try {
                    ContentCrawlDTO content = extractNovelData(href, mainDriver);
                    novels.add(content);
                } catch (TimeoutException e) {
                    System.err.println("TimeoutException 발생 - 다음 항목으로 진행: " + href);
                } catch (Exception e) {
                    System.err.println("예상치 못한 오류 발생 - 다음 항목으로 진행: " + href);
                    //e.printStackTrace();
                }
            }
        } catch (TimeoutException e) {
            throw new CrawlerException(CrawlerException.ExceptionMessage.TIMEOUT_EXCEPTION.getMessage(), e);
        } catch (InterruptedException e) {
            throw new CrawlerException(CrawlerException.ExceptionMessage.UNEXPECTED_EXCEPTION.getMessage(), e);
        }
        return novels;
    }

    private void setUpDrivers(WebDriver mainDriver) {
        mainDriver.get(KakaoNovelXPath.URL.getXpath());
        kakaoLogin.activateBot(mainDriver);
    }

    private void scrollPageToBottom(WebDriver driver) throws InterruptedException {
        JavascriptExecutor js = (JavascriptExecutor) driver;
        while (true) {
            Long lastHeight = (Long) js.executeScript("return document.body.scrollHeight");
            js.executeScript("window.scrollTo(0, document.body.scrollHeight)");
            Thread.sleep(1000);
            Long newHeight = (Long) js.executeScript("return document.body.scrollHeight");
            if (newHeight.equals(lastHeight)) {
                break;
            }
        }
    }

    private List<String> getKakaoPageData(WebDriver webDriver) {
        List<String> novelLinks = new ArrayList<>();

        WebDriverWait wait = new WebDriverWait(webDriver, Duration.ofSeconds(5));
        List<WebElement> elements = wait.until(driver -> driver.findElements(By.xpath(KakaoNovelXPath.LIST.getXpath())));

        for (WebElement el : elements) {
            String href = el.findElement(By.xpath(".//a")).getAttribute("href");
            novelLinks.add(href);
        }
        return novelLinks;
    }

    private void navigateToPage(WebDriver driver, String url) {
        driver.get(url);
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.urlToBe(url));
    }

    private ContentCrawlDTO extractNovelData(String detailHref, WebDriver webDriver) {
        String contentId = extractContentId(detailHref);
        String url = KakaoNovelXPath.DETAIL_URL.buildUrl(contentId);
        // 상세 페이지 이동
        navigateToPage(webDriver, url);

        // 상세 정보 추출
        String originalTitle = webDriver.findElement(By.xpath(KakaoNovelXPath.TITLE.getXpath())).getText();
        boolean isAdultContent = containsAdultTag(originalTitle); // 성인 여부 체크
        String title = extractTitle(originalTitle); // 실제 제목 정리
        String description = webDriver.findElement(By.xpath(KakaoNovelXPath.DESCRIPTION.getXpath())).getText();
        String coverImg = webDriver.findElement(By.xpath(KakaoNovelXPath.COVER_IMG.getXpath())).getAttribute("src");
        String genre = webDriver.findElement(By.xpath(KakaoNovelXPath.GENRE.getXpath())).getText();
        return new ContentCrawlDTO(title, description, coverImg, genre, isAdultContent, contentId);
    }

    private static boolean containsAdultTag(String title) {
        // [] 안의 내용을 추출해서 "19세"가 포함되어 있는지 확인
        Pattern pattern = Pattern.compile("\\[(.*?)\\]");
        Matcher matcher = pattern.matcher(title);

        while (matcher.find()) {
            if (matcher.group(1).contains("19세")) {
                return true; // 성인 콘텐츠로 판단
            }
        }
        return false;
    }

    private static String extractTitle(String title) {
        return title.replaceAll("\\[.*?\\]", "").trim();
    }

    private String extractContentId(String input) {
        Pattern pattern = Pattern.compile("\\d+");
        Matcher matcher = pattern.matcher(input);

        if (matcher.find()) {
            return matcher.group();
        } else {
            System.out.println("해당 작품 id를 찾을 수가 없습니다!");
            return null;
        }
    }
}

File: data_collector/crawl_collector/kakao/kakao_page_crawler.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from time import sleep
import logging
import re
from .kako_login import activate_bot
from ..xpaths.kakao_xpath import KakaoNovelXPath


class KakaoPageCrawler:

    def __init__(self, driver: webdriver.Chrome, xpath_class):
        self.driver = driver
        self.xpath_class = xpath_class
        self.driver.get(self.xpath_class.URL.value)
        activate_bot(driver)

    def crawl(self):
        novels = []
        try:
            self.scroll_page_to_bottom()

            novel_links = self.get_kakao_page_data()
            for href in novel_links:
                try:
                    content = self.extract_novel_data(href)
                    novels.append(content)
                except Exception as e:
                    logging.error(f"예상치 못한 오류 발생 - 다음 항목으로 진행: {href}")
                    # logging.exception(e)
        except TimeoutException as e:
            raise Exception("TimeoutException 발생", e)
        return novels

    def scroll_page_to_bottom(self):
        js = self.driver.execute_script
        while True:
            last_height = js("return document.body.scrollHeight")
            js("window.scrollTo(0, document.body.scrollHeight)")
            sleep(1)
            new_height = js("return document.body.scrollHeight")
            if new_height == last_height:
                break

    def get_kakao_page_data(self) -> list:
        novel_links = []

        wait = WebDriverWait(self.driver, 5)
        elements = wait.until(
            EC.presence_of_all_elements_located((By.XPATH, self.xpath_class.LIST.value))
        )

        for el in elements:
            href = el.find_element(By.XPATH, ".//a").get_attribute("href")
            novel_links.append(href)

        return novel_links

    def navigate_to_page(self, url: str):
        self.driver.get(url)

        wait = WebDriverWait(self.driver, 12)
        wait.until(EC.url_to_be(url))

    def extract_novel_data(self, detail_href: str) -> dict:
        content_id = self.extract_content_id(detail_href)
        url = self.xpath_class.DETAIL_URL.build_url(content_id)
        # 상세 페이지로 이동
        self.navigate_to_page(url)

        # 상세 정보 추출
        original_title = self.driver.find_element(
            By.XPATH, self.xpath_class.TITLE.value
        ).text

        is_adult_content = self.contains_adult_tag(original_title)  # 성인 여부 체크
        age_rating = 19 if is_adult_content else 12

        title = self.extract_title(original_title)  # 실제 제목 정리
        synopsis = self.driver.find_element(
            By.XPATH, self.xpath_class.DESCRIPTION.value
        ).text
        cover_img = self.driver.find_element(
            By.XPATH, self.xpath_class.COVER_IMG.value
        ).get_attribute("src")

        genre_raw = self.driver.find_element(
            By.XPATH, self.xpath_class.GENRE.value
        ).text
        genre = self.split_genre(genre_raw)

        return {
            "title": title,
            "synopsis": synopsis,
            "cover_img": cover_img,
            "genre": genre,
            "age_rating": age_rating,
            "content_id": content_id,
        }

    def contains_adult_tag(self, title: str) -> bool:
        pattern = r"\[(.*?)\]"
        matches = re.findall(pattern, title)
        for match in matches:
            if "19세" in match:
                return True
        return False

    def extract_title(self, title: str) -> str:
        return re.sub(r"\[.*?\]", "", title).strip()

    def extract_content_id(self, detail_href: str) -> str:
        # 숫자만 추출
        match = re.search(r"/content/(\d+)", detail_href)
        if match:
            return match.group(1)
        else:
            print("해당 작품 id를 찾을 수가 없습니다!")
            return None

    def split_genre(genre_raw: str) -> list[str]:
        if not genre_raw:
            return []
        # 여러 구분자 대응: · , / | 공백 포함
        return [g.strip() for g in re.split(r"[·/,|]", genre_raw) if g.strip()]
